promql/functions.go

```diff
@@ -406,17 +406,22 @@ func funcSortDesc(vals []parser.Value, args parser.Expressions, enh *EvalNodeHel
 
 // === sort_by_label(vector parser.ValueTypeVector, label parser.ValueTypeString...) (Vector, Annotations) ===
 func funcSortByLabel(vals []parser.Value, args parser.Expressions, enh *EvalNodeHelper) (Vector, annotations.Annotations) {
-	// In case the labels are the same, NaN should sort to the bottom, so take
-	// ascending sort with NaN first and reverse it.
-	var anno annotations.Annotations
-	vals[0], anno = funcSort(vals, args, enh)
+	// First, sort by the full label set. This ensures a consistent ordering in case sorting by the
+	// labels provided as arguments is not conclusive.
+	slices.SortFunc(vals[0].(Vector), func(a, b Sample) int {
+		return labels.Compare(a.Metric, b.Metric)
+	})
+
 	labels := stringSliceFromArgs(args[1:])
+	// Next, sort by the labels provided as arguments.
 	slices.SortFunc(vals[0].(Vector), func(a, b Sample) int {
```

Comment:
Doesn't this need to be `SortStableFunc` to achieve the intent of this PR?

(however I recommend you just move the comparison down here; sorting twice is unreasonably inefficient)

Reply:
Thank for the catch (and I blame the initial reviewer for not seeing it first :grin: ).

@suntala would you like to fix and refine this?

Reply:
Good point. I'll fix it. Thanks for the ping.

Reply:
Unfortunately, I won't be able to work on it this week. Something came up and I'll need to be OOO for several days. Definitely will make the fix though.

Reply:
Hi, did you make a start on the fix?  I would like to get it in to the 2.55 release.
I'm happy to look at it myself.

Reply:
https://github.com/prometheus/prometheus/pull/14985

Reply:
Hi, I was planning on it for Friday since I had to be OOO but don't want to block anything, especially since you have a fix already.